### PR TITLE
[UPDATE RESULTS COMPONENT]

### PR DESCRIPTION
**Pull request by RomeoPetryshyn**
Added histogram with groups reiting for current test. We you choose test the button appear to open it up as a modal window. Was done usign async/await feature in order not to have nested subscribes. Due to the situation that there are no endpoints on back-end in order to find out which groups passed current test - was needed to first of all find all students that passed current test, then to get group ids where those students studying, then work with those group ids to make needed amount of http requests in order to get the actual data for the histogram.
Also was needed to make a few changes in results component as it was needed to get testID in my component from this one
Everything appears to be working fine, added spinner while it's loading the data

### CODE REVIEW

File: src/app/admin/results/result-groups-raiting/result-groups-raiting.component.ts [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+  }
+
+  private getDataForChart() {
+    this.getScoreForChart();
+    this.getGroupNamesForChart();
+  }
+
+  private async getTestResultsList() {
+    const promise: any = this.resultsService.getRecordsByTestGroupDate(this.data.testID).toPromise();
+    return Promise.resolve(promise);
```
Comment (yurakhomitsky) [inline src/app/admin/results/result-groups-raiting/result-groups-raiting.component.ts]: Async function return a promise. What if we just write  `return promise` without `Promise.resolve(promise)`

File: src/app/admin/results/result-groups-raiting/result-groups-raiting.component.ts [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    const promise: any = this.apiService.getByEntityManager('Student', requestArray).toPromise();
+    return Promise.resolve(promise);
+  }
+
+  private async getGroupsResults() {
+    const studentsList = [...new Set(await this.getStudentsList())];
+    let groupIds: number[] = [];
+    studentsList.forEach((value: Student) => {
+      groupIds.push(value.group_id);
+    });
```
Comment (yurakhomitsky) [inline src/app/admin/results/result-groups-raiting/result-groups-raiting.component.ts]: Can we use map() here?

File: src/app/admin/results/result-groups-raiting/result-groups-raiting.component.ts [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+  private async getScoreForChart() {
+    const groupIds = [...new Set(await this.getGroupsResults())];
+    groupIds.forEach((value: number) => {
+      this.resultsService.getRecordsByTestGroupDate(this.data.testID, value).subscribe((data: Results[]) => {
+        let groupResult = 0;
+        data.forEach((item: Results) => {
+          groupResult += (Number(item.result) / Number(item.answers) * 100);
+        });
+        groupResult = Math.round(groupResult / data.length);
+        this.score.push(groupResult);
```
Comment (yurakhomitsky) [inline src/app/admin/results/result-groups-raiting/result-groups-raiting.component.ts]: Here we can use reduce function 
`const score = data.reduce((groupResult,item) => {
return groupResult += (Number(item.result)/Number(item.answers)* 100);
},0) `

File: src/app/admin/results/result-groups-raiting/result-groups-raiting.component.ts (hunk 1 above)
Comment (RomeoPetryshyn) [reply src/app/admin/results/result-groups-raiting/result-groups-raiting.component.ts]: It should work, but for better understanding and due to async/await example left like this, could be changed if it does metter

File: src/app/admin/results/result-groups-raiting/result-groups-raiting.component.ts (hunk 2 above)
Comment (RomeoPetryshyn) [reply src/app/admin/results/result-groups-raiting/result-groups-raiting.component.ts]: Yes, we can
Didn't thing that it does metter what to use in this case, can be easily changed if needed

File: src/app/admin/results/result-groups-raiting/result-groups-raiting.component.ts (hunk 3 above)
Comment (RomeoPetryshyn) [reply src/app/admin/results/result-groups-raiting/result-groups-raiting.component.ts]: Will check it

File: src/app/admin/results/result-groups-raiting/result-groups-raiting.component.ts (hunk 1 above)
Comment (yurakhomitsky) [reply src/app/admin/results/result-groups-raiting/result-groups-raiting.component.ts]: It works whether you write `Promise.resolve(promise)`  or  `return promise`. But second is preferable

Comment (RomeoPetryshyn) [reply src/app/admin/results/result-groups-raiting/result-groups-raiting.component.ts]: Got it

Comment (AndriyIf) [reply src/app/admin/results/result-groups-raiting/result-groups-raiting.component.ts]: return this.resultsService.getRecordsByTestGroupDate(this.data.testID).toPromise()

File: src/app/admin/results/result-groups-raiting/result-groups-raiting.component.ts (hunk 2 above)
Comment (AndriyIf) [reply src/app/admin/results/result-groups-raiting/result-groups-raiting.component.ts]: map is preferable for such cases

File: src/app/admin/results/result-groups-raiting/result-groups-raiting.component.ts (hunk 3 above)
Comment (AndriyIf) [reply src/app/admin/results/result-groups-raiting/result-groups-raiting.component.ts]: const score = data.reduce((groupResult,item) => { return groupResult + (Number(item.result)/Number(item.answers)* 100); },0)
Should be without `=`

File: src/app/admin/results/result-groups-raiting/result-groups-raiting.component.ts [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    this.getGroupNamesForChart();
+  }
+
+  private async getTestResultsList() {
+    const promise: any = this.resultsService.getRecordsByTestGroupDate(this.data.testID).toPromise();
+    return Promise.resolve(promise);
+  }
+
+  private async getStudentsList() {
+    const testResultsList = [...new Set(await this.getTestResultsList())];
```
Comment (AndriyIf) [inline src/app/admin/results/result-groups-raiting/result-groups-raiting.component.ts]: What is the reason of using Set?

File: src/app/admin/results/result-groups-raiting/result-groups-raiting.component.ts [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        this.score.push(groupResult);
+      });
+    });
+    this.chardReady = true;
+    this.isLoading = false;
+    this.chartData();
+  }
+
+  chartData() {
+    this.barChartLabels = this.groupNames;
```
Comment (AndriyIf) [inline src/app/admin/results/result-groups-raiting/result-groups-raiting.component.ts]: Looks weird. Can we use groupNames directly?

File: src/app/admin/results/result-groups-raiting/result-groups-raiting.component.ts (hunk 4 above)
Comment (RomeoPetryshyn) [reply src/app/admin/results/result-groups-raiting/result-groups-raiting.component.ts]: In order to leave just unique items in  array

File: src/app/admin/results/result-groups-raiting/result-groups-raiting.component.ts (hunk 5 above)
Comment (RomeoPetryshyn) [reply src/app/admin/results/result-groups-raiting/result-groups-raiting.component.ts]: In chartData()?

File: src/app/admin/results/result-groups-raiting/result-groups-raiting.component.ts (hunk 1 above)
Comment (RomeoPetryshyn) [reply src/app/admin/results/result-groups-raiting/result-groups-raiting.component.ts]: Got it, will fix today

File: src/app/admin/results/result-groups-raiting/result-groups-raiting.component.ts (hunk 5 above)
Comment (AndriyIf) [reply src/app/admin/results/result-groups-raiting/result-groups-raiting.component.ts]: In all component

**Comment by RomeoPetryshyn** [comment]
Everything was refactored as recommended

**Review by zikratyi — APPROVED** [review]

**Review by AndriyIf — APPROVED** [review]

**Pull request merged**